### PR TITLE
Student task status and study plan /wip

### PR DESCRIPTION
**Pull request by seejee**
Hi everyone,

@wicz and I are still working on this, but we'd like some preliminary feedback.

We've got tasks, their statuses, and a link to the study plan being displayed the student page. The course page provides a link to the student page.

In the next day or two, we're going to finish up marking tasks as complete from the UI, and showing/editing study plans.

Feedback is welcome. Thanks!

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -34,10 +29,27 @@ def login_path
     Rails.env.production? ? '/auth/github' : '/auth/developer'
   end
 
-  private
+  protected
```
Comment (jordanbyron) [inline app/controllers/application_controller.rb]: Why the change from `private` to `protected`?

**Comment by jordanbyron** [comment]
@geihsler & @wicz we need to talk about these tests. I'm really glad you are both writing tests, but a bunch of these are going to be thrown out once we restructure and finalize the views. That means I'll have failing tests anytime I make a minor change which isn't good.

What I think I'll do is after we merge this in, I'll create another pull request removing the unnecessary tests. Then in that pull request we can discuss why I removed them.

File: app/decorators/task_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
+class TaskDecorator < ApplicationDecorator
```
Comment (jordanbyron) [inline app/decorators/task_decorator.rb]: This decorator is a bit confusing, because I'd expect the `TaskDecorator` to decorate plain old `Task` objects, but in this case it doesn't. Instead it wraps a `Task` object and `StudentDecorator` object, which in the class is called a `participant`. This is what I'd call a `StudentTask` and isn't really a decorator. I think creating a proper `StudentTask` class is the right way to go here. That class should be able to build all the tasks for a student, much like `StudentDecorator#tasks` does. It should also handle returning the status of the task and contain the code from `CourseMembership` that has to do with tasks. How does that sound? It should help consolidate all the student task code which is spread across 2 or 3 classes.

Comment (seejee) [inline app/decorators/task_decorator.rb]: @jordanbyron good feedback.

The participant name comes from the fact that we were toying with renaming `CourseMembership` to `Participant` but decided not to. I'll rename that.

The `StudentTask` idea is an interesting one that, as you say, might consolidate the code from 2 or 3 classes. I see this class as a non-persisted model. Do you agree? Also, I think we'd then need a `StudentTaskDecorator` because `#status` and `#complete_task_link` are presentation logic. 

I'll play with this in our repo and ask you to review a pull request.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (seejee) [inline app/controllers/application_controller.rb]: fixed.

File: app/decorators/task_decorator.rb (hunk 2 above)
Comment (seejee) [inline app/decorators/task_decorator.rb]: @jordanbyron Please take a look at [this pull request](https://github.com/wicz/liskov/pull/6) where I made some of the changes that you recommended. Overall, I like it. What do you think?

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (wicz) [inline app/controllers/application_controller.rb]: ops, looks the former java programmer inside me is still alive. :wink:

**Comment by seejee** [comment]
@jordanbyron I'd like to talk about the tests as well. They've been very useful when developing these features, especially the integration tests. I'm curious to see which ones you think add to much friction to change. If a test is limiting change, than I'd like to find a way to make it a little more tolerant to change instead of deleting it. That being said, I'm not at all opposed to deleting tests that provide little or no value.

File: app/models/student_task.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    status != StudentTask::NOT_COMPLETE
+  end
+
+  def status
+    completed = get_completed_task
+    completed ? completed.description : NOT_COMPLETE
+  end
+
+  def complete(status)
+    #TODO: is there a better way to do an upsert? - cg
```
Comment (jordanbyron) [inline app/models/student_task.rb]: I think you're looking for:

``` ruby
course_membership.completed_tasks.find_or_create_by_task_id(:task_id => task_id, :description => status)
```

File: app/models/student_task.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    completed.save
+  end
+
+  private
+
+  def completed_tasks
+    @course_membership.completed_tasks
+  end
+
+  def get_completed_task
```
Comment (jordanbyron) [inline app/models/student_task.rb]: Let's rename this to just `completed_task`. The `get` part is implied and reminds me of java / C# :flushed:

File: app/views/student_tasks/edit.html.haml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+%p
+  =@student_task.description
+
+=form_tag @student_task.update_url, method: :put do
+  =label :student_task_, :status
+  =text_field :student_task, :status 
+  =submit_tag "Submit"
```
Comment (jordanbyron) [inline app/views/student_tasks/edit.html.haml]: This is a minor nit-pick, but add a space between the `=` operator and your ruby code. For example:

``` haml
= form_tag do
  = lable_tag "Blah"
```

File: app/decorators/student_decorator.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    course_membership.person.name
+  end
+
+  def tasks
+    course_membership.student_tasks.map {|st| StudentTaskDecorator.new(st)}
+  end
+
+  def study_plan
+    course_membership.study_plan || course_membership.create_study_plan
+  end
```
Comment (jordanbyron) [inline app/decorators/student_decorator.rb:14]: This is another place where `find_or_create_by` can be used

**Comment by jordanbyron** [comment]
This is coming along great guys. I really like the new `StudentTask` class. I made a few very small notes and once those are taken care of I think this is ready to merge. If either of you have any more questions just let me know. Great work!  :rocket:

File: app/decorators/student_decorator.rb (hunk 6 above)
Comment (seejee) [inline app/decorators/student_decorator.rb:14]: `study_plan` is a `:has_one` on `CourseMembership`. I looked through the API docs, and I couldn't find any way to `find_or_create` on a `:has_one` or would it be something like `StudyPlan.find_or_create_by_course_membership(course_membership)`?

Comment (seejee) [inline app/decorators/student_decorator.rb:14]: Nevermind. I got it.

**Comment by seejee** [comment]
Thanks for the review @jordanbyron! I believe I've addressed all of the concerns you had.

`StudentTask` really cleaned things up. Great suggestion!

**Comment by wicz** [comment]
Hey guys, I'd like to publicly thank @geihsler for all this work on this. You rock mate! :clap:

I'd like also to apologize for my absence. The last couple of days was like hell in work and with personal problems, so I couldn't help with anything. I hope tonight I'll be able to catch up with you and work on the remaining tasks.
